src/api/docker-compose.yml

```diff
@@ -46,3 +46,48 @@ services:
       - 'traefik.http.middlewares.strip_api_prefix.stripprefix.prefixes=/${API_VERSION}/image'
       - 'traefik.http.middlewares.strip_api_prefix.stripprefix.forceSlash=true'
       - 'traefik.http.routers.image.middlewares=strip_api_prefix'
+
+  redis:
+    image: redis:latest
+    ports:
+      - ${REDIS_PORT}:${REDIS_PORT}
```

Comment:
We don't want to expose Redis to the big-bad world.  We made this mistake once before, and got hacked, see https://docs.google.com/presentation/d/1xBYQyNEDJPS1JRlzw4JHAX1exuPTmTB-a950zHiU950/.  In Docker Compose, defining ports works like this:

```
ports:
  # Port definitions can take 1 or 2 values:
  # DOCKER_HOST_PORT: DOCKER_CONTAINER PORT
  6379: 6379
  # The above means: "use port 6379 on the HOST machine to expose port 6739 in the CONTAINER"
  # It has the effect of connecting the Redis server's port running in the container to the same port in
  # the host computer, effectively opening that port through to the internet, if the host is running on
  # the web (staging, production). 
```

This is what people tend to do when they work on development, and need to connect directly to some internal database or server.  By connecting the container's port to your local dev machine's port, you can access things running in the container.  It works great.  But in a production environment, you don't want to expose traffic via ports like this.  We want to route everything through Traefik, which can manage load-balancing, authentication requests, and container/port mappings.

For access to Redis and Elastic from the outside world, we want to use your service as our only direct means.  It shouldn't be possible to connect to https://telescope.cdot.systems:6739 for example.  Your service (and others that will talk to Redis/Elastic) will only be available on the outside via a set of well-defined HTTP endpoints, which can use JWT based authentication (see #1796).

The right way to do what you're doing here is to do something like you can see in these PRs:

1. https://github.com/Seneca-CDOT/telescope/pull/1840/files#diff-a3872c2e4d094f394e87eb51d5c934953e79319accea61231188218ba9b184b0  In this example I'm running a container for the Firebase emulator.  It uses all kinds of ports, but I only want to enable 2 of them: 4000 (Firestore) and 8088 (the Dashboard UI).  Notice that I say `- '4000'` (i.e., only 1 port).  I do this because I am defining the port that's being used in the container, but I am not exposing it to the outside world.  Instead, I do this to let Traefik map that internal port to an endpoint URL.  Going to http://ui.firebase.localhost will allow a developer to get to the container's port 4000, but I don't open port 4000 on the host machine:

```
      - 'traefik.http.routers.firebase_ui.rule=Host(`ui.firebase.localhost`)'
      - 'traefik.http.routers.firebase_ui.service=firebase_ui'
      - 'traefik.http.services.firebase_ui.loadbalancer.server.port=4000'
```

In the case of `8088` I do both a container port and a host port mapping.  I don't want to do this, but the Firebase UI expects to be able to use localhost:8088 in order to load JS/CSS and other static assets in the browser.  Because this is a development situation, I'm not that worried.  But FYI.

2. You can see what @raygervais is doing in https://github.com/Seneca-CDOT/telescope/pull/1802/files#diff-9b5638ed60397f4c9b63cc41ff257015946394e745dce431fda440513146f5de with the Search service and Elastic.  Here he defines the elastic port on the container, but doesn't open it to the host.

The docker-compose files are evolving quickly in various PRs that are in review right now, so don't worry too much about making this perfect.  I would advise that the easiest fix here is to drop the second port from each of these services so we don't expose anything.

I wanted to mention it to you so you'd know, since this isn't an obvious thing, and the cost of getting it wrong is much larger than you realize.

Reply:
Interesting, I forgot that the second port being there exposes the service to the web via the host machine. Noted